### PR TITLE
Check EXTENDED_READ, not CONFIGURE, on a context

### PR DESCRIPTION
**Pull request by jglick**
Needed for https://github.com/jenkinsci/workflow-cps-plugin/pull/12 to work well with multibranch, where the branch project is view-only but you may still want to work with configuration forms. The symptom is that _Credentials_ pulldowns are empty even when there are valid credentials available for the branch project.

@reviewbybees esp. @stephenc

### CODE REVIEW

**Comment by stephenc** [comment]
Hmmm we may need to think about this one. :bug: There are SECURITY implications involved in this change which we should at least discuss. Can you schedule some time on Monday?

File: src/main/java/hudson/plugins/git/UserRemoteConfig.java [hunk 1]
```diff
@@ -131,7 +131,7 @@ public FormValidation doCheckUrl(@AncestorInPath Item project,
                                          @QueryParameter String credentialsId,
                                          @QueryParameter String value) throws IOException, InterruptedException {
 
-            if (project == null || !project.hasPermission(Item.CONFIGURE)) {
+            if (project == null || !project.hasPermission(Item.EXTENDED_READ)) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: :bug: should be reverted, or switched to `CredentialsProvider.USE_ITEM`

Comment (jglick) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: (may be occasions where `USE_OWN` is valid, but pending JENKINS-31870 this does not work anyway: is checking `SYSTEM`)

**Comment by stephenc** [comment]
You may want to look into the changes I made for credentials 2.1.0 (see the changelog for the 2.1.0 release) https://wiki.jenkins-ci.org/display/JENKINS/Credentials+Plugin

> * Credentials providers need to be able to list credentials without retrieving the backing secret (issue #35306) - this change changes the recommended way to populate drop down select boxes for plugin authors. The old way still works but is no longer recommended, hence the minor version bump. An example of a new style implementation is as follows:
> ```
public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Job context, @QueryParameter String source, @QueryParameter String value) {
    if (context == null || !context.hasPermission(Item.CONFIGURE)) {
        // previously it was recommended to just return an empty ListBoxModel
        // now recommended to return a model with just the current value
        return new StandardUsernameListBoxModel().includeCurrentValue(value);
    }
    // previously it was recommended to use the withXXX methods providing the credentials instances directly
    // now recommended to populate the model using the includeXXX methods which call through to
    // CredentialsProvider.listCredentials and to ensure that the current value is always present using
    // includeCurrentValue
    return new StandardUsernameListBoxModel()
            .includeEmptySelection()
            .includeAs(Tasks.getAuthenticationOf(context), context, StandardUsernameCredentials.class,
                URIRequirementBuilder.fromUri(source).build())
            .includeCurrentValue(value);
}
```
>    The main changes in this sample are:
>
>    * Adding a QueryParameter for the current value. We can then ensure that the current value is always available by using the includeCurrentValue helper method
>    * Using includeAs to add the credentials using CredentialsProvider.listCredentials rather than the old pattern whereby the credentials would be retrieved directly and then converted into display name & id
>    * Now recommended to use Tasks.getAuthenticationOf(job) to get the authentication that the job will run as. When issue #35081 is resolved this will become more important


Specifically I am thinking that `new StandardUsernameListBoxModel().includeCurrentValue(value)` may be how you want to handel EXTENDED_READ

**Comment by jglick** [comment]
All seems good, but how does it relate to this PR other than affecting nearby code?

> `new StandardUsernameListBoxModel().includeCurrentValue(value)` may be how you want to handle `EXTENDED_READ`

I do not think so. To reiterate from the PR description:

> The symptom is that _Credentials_ pulldowns are empty even when there are valid credentials available for the branch project.

Here the `value` would of course be null since it is a fresh configuration form in _Snippet Generator_, so `includeCurrentValue` would be a no-op. We want to display all contextually valid credentials (`id` + `displayName` at least, if using a 2.1.0 dep).

**Comment by stephenc** [comment]
Well if you want to display contextually valid ones be sure to pay attention to the job's default authentication too.

**Comment by jglick** [comment]
> be sure to pay attention to the job's default authentication too

Might need to do something there, but seems a topic for a separate PR.

**Comment by jglick** [comment]
> seems a topic for a separate PR

Probably just means resolving merge conflicts with #412, right?

**Comment by stephenc** [comment]
FYI: If the user does not have `Item/Configure` then likely they will not have `Credentials/UseItem` so they will be unable to see the credentials available anyway

**Comment by jglick** [comment]
> If the user does not have `Item/Configure` then likely they will not have `Credentials/UseItem`

Not true in the case of branch projects (and multibranch children of organization folders), where `Item.CONFIGURE` is specifically prohibited to all (non-`SYSTEM`) users, yet permissions normally `impliedBy` `Item.CONFIGURE` are retained.

File: src/main/java/hudson/plugins/git/UserRemoteConfig.java [hunk 2]
```diff
@@ -134,7 +133,9 @@ public FormValidation doCheckUrl(@AncestorInPath Item item,
                                          @QueryParameter String credentialsId,
                                          @QueryParameter String value) throws IOException, InterruptedException {
 
-            if (item == null || !item.hasPermission(Item.CONFIGURE)) {
+            // Normally this permission is hidden and implied by Item.CONFIGURE, so from a view-only form you will not be able to use this check.
+            // (Under certain circumstances being granted only USE_OWN might suffice, though this presumes a fix of JENKINS-31870.)
```
Comment (stephenc) [inline src/main/java/hudson/plugins/git/UserRemoteConfig.java]: :ant: I'd like to see a `//TODO` here to address the `USE_OWN` scenario

**Comment by stephenc** [comment]
:bee:

**Comment by stephenc** [comment]
:bee:

**Comment by MarkEWaite** [comment]
Should this be combined with #412 into a single release of the git plugin or is it needed sooner than I'm likely to complete the review and inclusion of #412 ?

I don't feel especially qualified to evaluate this other than through interactive testing, and I'm suspicious that my interactive testing is unlikely to expose the many interesting ways that permissions can interact with job definitions.  I'm relying on @stephenc and @jglick evaluation, and I assume the :bee: icon from @stephenc means that he believes this is ready for interactive test and merge.

**Comment by jglick** [comment]
```
org.apache.maven.artifact.InvalidArtifactRTException: For artifact {null:null:null:jar}: The groupId cannot be empty.
	at org.apache.maven.artifact.DefaultArtifact.validateIdentity(DefaultArtifact.java:130)
	at org.apache.maven.artifact.DefaultArtifact.<init>(DefaultArtifact.java:123)
	at org.apache.maven.artifact.factory.DefaultArtifactFactory.createArtifact(DefaultArtifactFactory.java:157)
	at org.apache.maven.artifact.factory.DefaultArtifactFactory.createArtifact(DefaultArtifactFactory.java:117)
	at org.apache.maven.artifact.factory.DefaultArtifactFactory.createArtifact(DefaultArtifactFactory.java:111)
	at org.apache.maven.artifact.factory.DefaultArtifactFactory.createArtifact(DefaultArtifactFactory.java:39)
	at org.apache.maven.repository.legacy.LegacyRepositorySystem.createArtifact(LegacyRepositorySystem.java:119)
	at org.apache.maven.project.DefaultProjectBuilder.initProject(DefaultProjectBuilder.java:604)
	at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:160)
	at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:312)
	at org.apache.maven.project.DefaultProjectBuilder.build(DefaultProjectBuilder.java:270)
	at org.apache.maven.project.artifact.MavenMetadataSource.retrieveRelocatedProject(MavenMetadataSource.java:578)
	at org.apache.maven.project.artifact.MavenMetadataSource.retrieve(MavenMetadataSource.java:191)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:532)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:584)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:584)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.recurse(DefaultLegacyArtifactCollector.java:584)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.collect(DefaultLegacyArtifactCollector.java:144)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.collect(DefaultLegacyArtifactCollector.java:100)
	at org.apache.maven.repository.legacy.resolver.DefaultLegacyArtifactCollector.collect(DefaultLegacyArtifactCollector.java:782)
	at org.apache.maven.shared.dependency.tree.DefaultDependencyTreeBuilder.buildDependencyTree(DefaultDependencyTreeBuilder.java:118)
	at org.apache.maven.shared.dependency.tree.DefaultDependencyTreeBuilder.buildDependencyTree(DefaultDependencyTreeBuilder.java:163)
	at org.apache.maven.plugins.enforcer.AbstractResolveDependencies.getDependenciesToCheck(AbstractResolveDependencies.java:92)
	at org.apache.maven.plugins.enforcer.AbstractResolveDependencies.execute(AbstractResolveDependencies.java:75)
	at org.apache.maven.plugins.enforcer.EnforceMojo.execute(EnforceMojo.java:193)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
```

is the same as an error I have been getting in `junit`, I suspect due to a problem in Artifactory.

**Comment by jglick** [comment]
> Should this be combined with #412 into a single release

I believe they are independent.

> I'm relying on @stephenc and @jglick evaluation, and I assume the bee icon from @stephenc means that he believes this is ready for interactive test and merge.

Yes—means he does not see anything wrong with this (whether or not he specifically advocates it).

**Comment by jglick** [comment]
By the way the way to see this effect is to create a multibranch Pipeline project (using a Git branch source, say), and then once a branch project (like `master`) has been created, click _Pipeline Syntax_ and under `checkout` select `GitSCM` and look in the credentials pulldown. Prior to fix it will be empty; after the fix you should see available credentials, assuming the _View Configuration_ link is visible to you.

**Pull request merged**